v2/rest/client.go

```diff
@@ -0,0 +1,310 @@
+package gohierarchyclient
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/SKF/go-hierarchy-client/v2/rest/models"
+
+	rest "github.com/SKF/go-rest-utility/client"
+	"github.com/SKF/go-utility/v2/stages"
+	"github.com/SKF/go-utility/v2/uuid"
+)
+
+type TreeFilter struct {
+	Depth         int
+	Limit         int
+	Offset        int
+	MetadataKey   string
+	MetadataValue string
+	NodeTypes     []string
+	ModifiedAfter string
+}
+
+type HierarchyClient interface {
+	GetNode(ctx context.Context, id uuid.UUID) (models.GetNodeResponse, error)
+	CreateNode(ctx context.Context, node models.CreateNodeRequest) (models.CreateNodeResponse, error)
+	UpdateNode(ctx context.Context, id uuid.UUID, node models.UpdateNodeRequest) (models.UpdateNodeResponse, error)
+	DeleteNode(ctx context.Context, id uuid.UUID) error
+	DuplicateNode(ctx context.Context, source uuid.UUID, destination uuid.UUID, suffix string) (models.DuplicateNodeResponse, error)
+
+	GetAncestors(ctx context.Context, id uuid.UUID, height int, nodeTypes ...string) (models.GetAncestorsResponse, error)
+	GetCompany(ctx context.Context, id uuid.UUID) (models.GetCompanyResponse, error)
+	GetSubtree(ctx context.Context, id uuid.UUID, filter TreeFilter, continuationToken string) (models.GetSubtreeResponse, error)
+	GetSubtreeCount(ctx context.Context, id uuid.UUID, nodeTypes ...string) (models.GetSubtreeCountResponse, error)
+
+	LockNode(ctx context.Context, id uuid.UUID, recursive bool) error
+	UnlockNode(ctx context.Context, id uuid.UUID, recursive bool) error
+
+	GetOrigins(ctx context.Context, provider, continuationToken string, limit int) (models.GetOriginsResponse, error)
+	GetOriginsByType(ctx context.Context, provider, originType, continuationToken string, limit int) (models.GetOriginsResponse, error)
+	GetProviderNodeIDs(ctx context.Context, provider, continuationToken string, limit int) (models.GetNodesByPartialOriginResponse, error)
+	GetProviderNodeIDsByType(ctx context.Context, provider, originType, continuationToken string, limit int) (models.GetNodesByPartialOriginResponse, error)
+	GetOriginNodeID(ctx context.Context, origin models.Origin) (models.GetNodeByOriginResponse, error)
+}
+
+type client struct {
+	*rest.Client
+	clientID string
+}
+
+func WithStage(stage string) rest.Option {
+	if stage == stages.StageProd {
+		return rest.WithBaseURL("https://api.hierarchy.enlight.skf.com")
+	}
+
+	return rest.WithBaseURL(fmt.Sprintf("https://api.%s.hierarchy.enlight.skf.com", stage))
+}
+
+func NewClient(clientID string, opts ...rest.Option) HierarchyClient {
+	restClient := rest.NewClient(
+		append([]rest.Option{
+			// Defaults to production stage if no option is supplied
+			WithStage(stages.StageProd),
+		}, opts...)...,
+	)
+
+	return &client{Client: restClient, clientID: clientID}
+}
+
+func (c *client) GetNode(ctx context.Context, id uuid.UUID) (models.GetNodeResponse, error) {
+	request := rest.Get("v2/nodes/{node}").
```

Comment:
I think v2 should be part of the baseURL?

Reply:
baseURL removec everything after the top domain, no?

Reply:
I dont think so. If you include an leading `/` it will remove any path elements but not if you include `v2` in the baseURL and have,
```go
rest.Get("nodes/{node}")
```

Reply:
Doesn't seem like it
`something := rest.WithBaseURL("https://api.hierarchy.enlight.skf.com/v2")
	client := rest.NewClient(something)
	req := rest.Get("nodes/123")
	_, err := client.Do(context.Background(), req)

	fmt.Println("lol ", client.BaseURL)
	fmt.Println("lol2 ", err)`

prints 
`lol  https://api.hierarchy.enlight.skf.com/v2
lol2  got 401 for https://api.hierarchy.enlight.skf.com/nodes/123: {"error": {"message": "unauthorized"}}: unauthorized`

Reply:
Ah you mean trailing slash :)

Reply:
Ah, you might need to include a trailing slash in the base URL,
```go
func TestExpandURLWithPath(t *testing.T) {
	baseURL := urlMustParse("https://example.com/path/")
	request := Get("endpoint")

	url, err := request.ExpandURL(baseURL)

	require.NoError(t, err)
	require.Equal(t, "https://example.com/path/endpoint", url.String())
}
```